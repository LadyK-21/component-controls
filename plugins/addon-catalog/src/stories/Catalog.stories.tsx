import React from 'react';
import { Document, Example } from '@component-controls/core';
import { BlockContextProvider, store } from '@component-controls/blocks';
import { Catalog } from '../Catalog';

export default {
  title: 'Plugins/AddonCatalog/Catalog',
  component: Catalog,
  category: 'Components',
} as Document;

export const overview: Example = () => {
  return (
    <BlockContextProvider
      storyId="blocks-core-story-plain--controls"
      store={store}
    >
      <Catalog />
    </BlockContextProvider>
  );
};
