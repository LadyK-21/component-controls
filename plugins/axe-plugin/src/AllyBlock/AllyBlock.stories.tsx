import React from 'react';
import { Document, Example } from '@component-controls/core';
import { BlockContextProvider, store } from '@component-controls/blocks';
import { AllyBlock } from '../index';

export default {
  title: 'Plugins/AllyBlock',
  component: AllyBlock,
  category: 'Testing',
} as Document;

export const overview: Example = () => {
  return (
    <BlockContextProvider
      storyId="blocks-core-story-plain--controls"
      store={store}
    >
      <AllyBlock id="." />
    </BlockContextProvider>
  );
};
