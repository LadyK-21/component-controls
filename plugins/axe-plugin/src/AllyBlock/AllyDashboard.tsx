import React, { FC, useContext } from 'react';
import { AxeResults } from 'axe-core';
import { Chart } from 'react-google-charts';
import { Flex, Heading, Grid, Label, Checkbox } from 'theme-ui';
import {
  AxeContext,
  useTaggedList,
  SelectionContext,
  useIsTagSelected,
  tagSelectedList,
} from '../state/context';

type StatsStatus = 'passes' | 'violations';

const TagSelectionCheckbox: FC<{ tag: string }> = ({ tag }) => {
  const isSelected = useIsTagSelected(tag);
  const { selection, setSelection } = useContext(SelectionContext);
  const tagged = useTaggedList();
  const toggleTagSelected = () => {
    if (tagged[tag]) {
      const newSelection = tagSelectedList(selection, tagged[tag], false);
      if (!isSelected) {
        Array.prototype.push.apply(newSelection, tagged[tag]);
      }
      setSelection(newSelection);
    }
  };

  return (
    <Flex
      css={{
        flexDirection: 'column',
        alignItems: 'center',
      }}
    >
      <Label>
        <Checkbox onChange={() => toggleTagSelected()} checked={isSelected} />
        {isSelected ? 'hide outlines' : 'show errors'}
      </Label>
    </Flex>
  );
};

interface Stats {
  [cat: string]: {
    [key in StatsStatus]: number;
  };
}
const collectStats = (
  results: Pick<AxeResults, 'violations' | 'passes'>,
  status: StatsStatus,
  initial: Stats,
) => {
  const success = results[status].reduce((acc, row) => {
    const cats = row.tags.filter(tag => tag.split('.')[0] === 'cat');
    return cats.reduce((a, cat) => {
      const c = cat.split('.')[1];
      if (a[c] === undefined) {
        a[c] = {
          passes: 0,
          violations: 0,
        };
      }
      return { ...a, [c]: { ...a[c], [status]: a[c][status] + 1 } };
    }, acc);
  }, initial);
  return success;
};
export const AllyDashboard: FC = () => {
  const { results } = useContext(AxeContext);
  const stats = collectStats(
    results,
    'violations',
    collectStats(results, 'passes', {}),
  );
  return (
    <Grid gap={2} columns={[3, '1fr 1fr 1fr']}>
      {Object.keys(stats)
        .sort()
        .map(key => {
          const value = stats[key];
          const rate = Math.trunc(
            (100 * (value.passes - value.violations)) /
              (value.passes + value.violations),
          );
          return (
            <Flex
              key={key}
              css={{
                height: 260,
                alignItems: 'center',
                flexDirection: 'column',
              }}
            >
              <Heading as="h4">{key}</Heading>
              <Chart
                chartType="Gauge"
                options={{
                  redFrom: -100,
                  redTo: -50,
                  yellowFrom: -50,
                  yellowTo: 50,
                  greenFrom: 50,
                  greenTo: 100,
                  majorTicks: [],
                  minorTicks: 0,
                  min: -100,
                }}
                data={[
                  ['Label', 'Value'],
                  [
                    `${
                      value.violations
                        ? `-${value.violations}${
                            value.passes ? `/+${value.passes}` : ''
                          }`
                        : `+${value.passes}`
                    } rules`,
                    rate,
                  ],
                ]}
              />
              {value.violations ? (
                <TagSelectionCheckbox tag={`cat.${key}`} />
              ) : null}
            </Flex>
          );
        })}
    </Grid>
  );
};
