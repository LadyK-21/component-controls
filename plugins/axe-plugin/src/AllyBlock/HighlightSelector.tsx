/** @jsx jsx */
import { FC, useContext, useMemo } from 'react';
import { jsx, ThemeUICSSObject, Box, BoxProps } from 'theme-ui';
import { SelectionContext, Selection } from '../state/context';

export interface StyledContainerProps {
  selection: Selection;
}
const HighlightNode: FC<StyledContainerProps & BoxProps> = ({
  selection,
  ...rest
}) => {
  const styles: ThemeUICSSObject | undefined = useMemo(() => {
    const selectors = selection.map(c => {
      const style = c.split('.story-render-container');
      let selector: string;
      if (style.length >= 2) {
        selector = `.story-render-container${
          style[style.length - 1] ? style[style.length - 1] : ':first-child'
        }`;
      } else if (c === ':root') {
        selector = '.story-render-container > * ';
      } else {
        selector = c;
      }
      return `${selector}`;
    });
    return selectors.length
      ? {
          [`${selectors.join(', ')}`]: {
            outline: '3px dotted red',
            outlineOffset: '3px',
          },
        }
      : undefined;
  }, [selection]);
  return <Box sx={styles} {...rest} />;
};

export const HighlightSelector: FC = ({ children }) => {
  const { selection } = useContext(SelectionContext);
  return <HighlightNode selection={selection}>{children}</HighlightNode>;
};
