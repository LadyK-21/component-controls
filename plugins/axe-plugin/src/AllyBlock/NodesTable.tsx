/* eslint-disable react/display-name */
/** @jsx jsx */
import { FC, useMemo, useContext } from 'react';
import { jsx, Flex, Box, Label, Checkbox } from 'theme-ui';
import { NodeResult } from 'axe-core';
import {
  SyntaxHighlighter,
  Table,
  Column,
  Tag,
} from '@component-controls/components';
import { SelectionContext, tagSelectedList, trimNode } from '../state/context';

export interface NodesTableProps {
  /**
   * list of nodes fromthe results
   */
  nodes?: NodeResult[];
  /**
   * if true, will hide error columns such as impact
   */
  hideErrorColumns?: boolean;
}

const SelectionCheckbox: FC<{ target: string[] }> = ({ target }) => {
  const { isSelected, selection, setSelection } = useContext(SelectionContext);

  const checked = isSelected(target);
  const toggleSelection = (selector: string[]) => {
    const included = tagSelectedList(selection, selector, true);
    if (included.length) {
      setSelection(tagSelectedList(selection, selector, false));
    } else {
      setSelection([...selection, ...selector]);
    }
  };
  return (
    <Label>
      <Checkbox onChange={() => toggleSelection(target)} checked={checked} />
    </Label>
  );
};
export const NodesTable: FC<NodesTableProps> = ({
  nodes,
  hideErrorColumns,
}) => {
  const columns: Column<NodeResult>[] = useMemo(
    () => [
      {
        Header: '',
        accessor: 'target',
        width: 80,
        Cell: ({
          row: {
            original: { target },
          },
        }: {
          row: { original: NodeResult };
        }) => <SelectionCheckbox target={target} />,
      },
      {
        Header: 'summary',
        accessor: 'failureSummary',
      },
      {
        Header: 'html',
        accessor: 'html',
        Cell: ({ value }: { value: string }) => {
          return (
            <Flex
              sx={{
                '& .token-line': {
                  whiteSpace: 'break-spaces',
                },
              }}
            >
              <SyntaxHighlighter language="markup">{value}</SyntaxHighlighter>
            </Flex>
          );
        },
      },
      {
        Header: 'target',
        accessor: 'target',
        Cell: ({ value }: { value: string[] }) => {
          return (
            <Flex
              css={{
                flexWrap: 'wrap',
              }}
            >
              {value &&
                value.map(target => (
                  <Tag
                    key={`${target}`}
                    color="lightgrey"
                    variant="tag.rightmargin"
                  >
                    {trimNode(target)}
                  </Tag>
                ))}
            </Flex>
          );
        },
      },
    ],
    [],
  );
  return (
    <Box
      sx={{
        backgroundColor: 'shadow',
      }}
    >
      <Table<NodeResult>
        data={nodes}
        columns={columns}
        hiddenColumns={
          hideErrorColumns ? ['failureSummary', 'targets'] : ['targets']
        }
      />
    </Box>
  );
};
