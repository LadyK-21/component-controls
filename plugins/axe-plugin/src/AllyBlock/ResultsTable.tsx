/* eslint-disable react/display-name */
/** @jsx jsx */
import { FC, useMemo, useCallback, useContext } from 'react';
import { jsx, Flex, Box, Text } from 'theme-ui';
import {
  ChevronRightIcon,
  ChevronDownIcon,
  TriangleRightIcon,
  TriangleDownIcon,
  AlertIcon,
  StopIcon,
  Icon,
  InfoIcon,
  IssueOpenedIcon,
} from '@primer/octicons-react';
import { Result, ImpactValue } from 'axe-core';
import {
  Table,
  Column,
  ExternalLink,
  Tag,
} from '@component-controls/components';
import { AxeContext } from '../state/context';
import { NodesTable } from './NodesTable';

const impactColors: {
  [key in 'minor' | 'moderate' | 'serious' | 'critical']: {
    color: string;
    icon: Icon;
  };
} = {
  minor: {
    color: '#2196f3',
    icon: InfoIcon,
  },
  moderate: {
    color: '#f57c00',
    icon: IssueOpenedIcon,
  },
  serious: {
    color: '#e57373',
    icon: AlertIcon,
  },
  critical: {
    color: '#dc004e',
    icon: StopIcon,
  },
};
export interface ResultsTableProps {
  /**
   * array of scan results
   */
  results?: Result[];
  /**
   * if true, will hide error columns such as impact
   */
  hideErrorColumns?: boolean;
}

const ResultsTable: FC<ResultsTableProps> = ({ results, hideErrorColumns }) => {
  const renderRowSubComponent = useCallback(
    ({ row }) => (
      <NodesTable
        nodes={row.original.nodes}
        hideErrorColumns={hideErrorColumns}
      />
    ),
    [hideErrorColumns],
  );
  const table = useMemo(() => {
    const columns: Column<Result>[] = [
      {
        // Build our expander column
        id: 'expander', // Make sure it has an ID
        Header: ({ getToggleAllRowsExpandedProps, isAllRowsExpanded }: any) => (
          <span {...getToggleAllRowsExpandedProps()}>
            {isAllRowsExpanded ? <TriangleDownIcon /> : <TriangleRightIcon />}
          </span>
        ),
        width: 50,
        Cell: ({ row }: any) => (
          <Flex
            {...row.getToggleRowExpandedProps()}
            sx={{
              pl: 2,
            }}
          >
            {row.isExpanded ? <ChevronDownIcon /> : <ChevronRightIcon />}
          </Flex>
        ),
      },
      {
        Header: 'id',
        accessor: 'id',
        Cell: ({
          row: {
            original: { helpUrl, id },
          },
        }: any) => {
          const el = <Box css={{ whiteSpace: 'nowrap' }}>{id}</Box>;
          if (!helpUrl) {
            return el;
          }
          return <ExternalLink href={helpUrl}>{el}</ExternalLink>;
        },
      },
      {
        Header: 'impact',
        accessor: 'impact',
        Cell: ({ value }: { value: ImpactValue }) => {
          const impact = value ? impactColors[value] : undefined;
          return (
            <Flex
              sx={{
                alignItems: 'center',
                flexDirection: 'row',
                color: impact ? impact.color : undefined,
              }}
            >
              {impact && <impact.icon />}
              <Text
                sx={{
                  pl: impact ? 2 : 0,
                }}
              >
                {value}
              </Text>
            </Flex>
          );
        },
      },
      {
        Header: 'description',
        accessor: 'description',
      },
      {
        Header: 'tags',
        accessor: 'tags',
        Cell: ({ value }: { value: string[] }) => (
          <Flex
            css={{
              flexWrap: 'wrap',
            }}
          >
            {value &&
              value.map(tag => (
                <Tag key={`${tag}`} color="lightgrey" variant="tag.rightmargin">
                  {tag}
                </Tag>
              ))}
          </Flex>
        ),
      },
    ];
    return (
      <Table<Result>
        data={results || []}
        columns={columns}
        hiddenColumns={hideErrorColumns ? ['impact'] : undefined}
        renderRowSubComponent={renderRowSubComponent}
      />
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [results?.length, hideErrorColumns, renderRowSubComponent]);

  return table;
};

export const ViolationsTable: FC = () => {
  const {
    results: { violations },
  } = useContext(AxeContext);

  return <ResultsTable results={violations} />;
};

export const PassesTable: FC = () => {
  const {
    results: { passes },
  } = useContext(AxeContext);
  return <ResultsTable results={passes} hideErrorColumns={true} />;
};

export const IncompleteTable: FC = () => {
  const {
    results: { incomplete },
  } = useContext(AxeContext);

  return <ResultsTable results={incomplete} />;
};
