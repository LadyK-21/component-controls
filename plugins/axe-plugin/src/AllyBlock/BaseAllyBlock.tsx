/* eslint-disable react/display-name */
import React, { FC, useContext, useState, useEffect, useCallback } from 'react';
import { Spec } from 'axe-core';

import { ActionContainer } from '@component-controls/components';
import { ViolationsTable, PassesTable, IncompleteTable } from './ResultsTable';
import { HighlightSelector } from './HighlightSelector';
import { AxeContext } from '../state/context';

/**
 * options
 */
export interface BaseAllyBlockProps {
  options?: Spec;
}

type ItemsState = 'violations' | 'passes' | 'incomplete';

const PanelState: Record<ItemsState, FC> = {
  violations: ViolationsTable,
  passes: PassesTable,
  incomplete: IncompleteTable,
};
/**
 * Displays the [axe](https://github.com/dequelabs/axe-core) ally test results
 */
export const BaseAllyBlock: FC<BaseAllyBlockProps> = ({ children }) => {
  const {
    results: { violations, passes, incomplete },
  } = useContext(AxeContext);
  const defaultState = useCallback(
    () =>
      violations.length
        ? 'violations'
        : incomplete.length
        ? 'incomplete'
        : 'passes',
    [incomplete.length, violations.length],
  );
  const [state, setState] = useState<ItemsState>(defaultState());
  useEffect(() => {
    setState(defaultState());
  }, [defaultState]);
  const numPanels =
    (violations.length ? 1 : 0) +
    (incomplete.length ? 1 : 0) +
    (passes.length ? 1 : 0);
  const nextState = (): ItemsState => {
    switch (state) {
      case 'violations':
        if (passes.length) {
          return 'passes';
        }
        return 'incomplete';
      case 'passes':
        if (incomplete.length) {
          return 'incomplete';
        }
        return 'violations';
      case 'incomplete':
        if (violations.length) {
          return 'violations';
        }
        return 'passes';
    }
  };

  const Panel = PanelState[state];
  return (
    <ActionContainer
      actions={
        numPanels > 1
          ? [
              {
                node: `view ${nextState()}`,
                id: 'panel',
                'aria-label': `display the accessibility ${nextState()}`,
                onClick: () => setState(nextState()),
              },
            ]
          : undefined
      }
    >
      <HighlightSelector>{children}</HighlightSelector>
      {numPanels > 0 && <Panel />}
    </ActionContainer>
  );
};
