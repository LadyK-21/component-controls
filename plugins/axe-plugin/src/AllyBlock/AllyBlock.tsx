/* eslint-disable react/display-name */
import React, { FC, useRef, useContext, useState, useEffect } from 'react';
import {
  run as runAxe,
  configure as configureAxe,
  reset,
  cleanup,
} from 'axe-core';

import { useStory, StoryInputProps } from '@component-controls/store';
import { resetTabCounter } from '@component-controls/components';

import {
  StoryBlockContainer,
  StoryBlockContainerProps,
  Story,
} from '@component-controls/blocks';
import { Spec } from 'axe-core';

import { BaseAllyBlock } from './BaseAllyBlock';
import {
  AxeContextProvider,
  AxeSetContext,
  SelectionContextProvider,
} from '../state/context';

interface AllyBlockOwmProps {
  axeOptions?: Spec;
}

export type AllyBlockProps = AllyBlockOwmProps &
  StoryInputProps &
  StoryBlockContainerProps;

const RenderStory: FC<AllyBlockOwmProps & { storyId?: string }> = ({
  axeOptions,
  storyId,
}) => {
  const storyRef = useRef<HTMLDivElement>(null);
  const { setResults } = useContext(AxeSetContext);
  const [mounted, setMounted] = useState(true);
  const isRunning = useRef(false);
  useEffect(() => {
    setMounted(true);
    return () => setMounted(false);
  }, []);
  const collectResults = () => {
    const canvas = storyRef.current?.firstChild;
    if (canvas && isRunning.current === false) {
      isRunning.current = true;
      reset();
      configureAxe(axeOptions || {});
      resetTabCounter();
      runAxe(canvas)
        .then(results => {
          if (mounted) {
            const { passes, violations, incomplete } = results;
            setResults({ passes, violations, incomplete });
            setTimeout(() => (isRunning.current = false), 1000);
          }
        })
        .catch(e => {
          console.error('error running axe-core', e);
          isRunning.current = false;
        });
    }
  };
  const onRender = () => {
    try {
      collectResults();
    } catch (e) {}
  };
  return (
    <Story key={storyId} id={storyId} ref={storyRef} onRender={onRender} />
  );
};

/**
 * Story block container that displays displays the [axe](https://github.com/dequelabs/axe-core) ally test results
 */
export const AllyBlock: FC<AllyBlockProps> = ({
  axeOptions,
  id,
  name,
  ...props
}) => {
  const story = useStory({ id, name });
  useEffect(() => {
    return () => {
      cleanup();
    };
  }, []);
  return story ? (
    <StoryBlockContainer story={story} {...props}>
      <AxeContextProvider>
        <SelectionContextProvider>
          <BaseAllyBlock options={axeOptions}>
            <RenderStory storyId={story.id} axeOptions={axeOptions} />
          </BaseAllyBlock>
        </SelectionContextProvider>
      </AxeContextProvider>
    </StoryBlockContainer>
  ) : null;
};
