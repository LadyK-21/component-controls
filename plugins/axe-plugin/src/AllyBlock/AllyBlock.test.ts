import * as path from 'path';

import { loadConfigurations } from '@component-controls/config';
import { renderDocument } from '@component-controls/test-renderers';
import { render, act } from '@testing-library/react';
import { renderErr } from '@component-controls/test-renderers';

import * as examples from './AllyBlock.stories';

describe('AllyBlock', () => {
  const configPath = path.resolve(__dirname, '../../.config');
  const config = loadConfigurations(configPath);
  let renderedExamples: ReturnType<typeof renderDocument> = [];
  act(() => {
    renderedExamples = renderDocument(examples, config);
  });
  if (!renderedExamples) {
    renderErr();
    return;
  }
  renderedExamples.forEach(({ name, rendered }) => {
    describe(name, () => {
      it('snapshot', () => {
        const { asFragment } = render(rendered);
        expect(asFragment()).toMatchSnapshot();
      });
    });
  });
});
