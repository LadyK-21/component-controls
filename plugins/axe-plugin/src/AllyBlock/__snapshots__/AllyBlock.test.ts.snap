// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllyBlock overview snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  margin-top: 32px;
  margin-bottom: 32px;
  width: 100%;
  scroll-margin-top: 5rem;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  width: 100%;
  scroll-margin-top: 5rem;
}

.emotion-2 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2:hover a {
  visibility: visible;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.1);
  border: 1px solid rgba(0, 0, 0, 0.1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.emotion-8 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  padding-left: 32px;
  padding-right: 32px;
  padding-top: 16px;
  padding-bottom: 16px;
}

.emotion-9 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  all: unset;
}

.emotion-10 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  color: #5c6ac4;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
      id=""
    >
      <div
        class="emotion-2"
      />
    </div>
    <div
      aria-hidden="false"
      class="rah-static rah-static--height-auto"
      style="height: auto; overflow: visible;"
    >
      <div>
        <div>
          <div
            class="emotion-3"
          >
            <div
              class="emotion-4"
            >
              <div
                class="emotion-0"
                data-testid="story"
              >
                <div
                  class="emotion-1"
                  id=""
                >
                  <div
                    class="emotion-2"
                  />
                </div>
                <div
                  aria-hidden="false"
                  class="rah-static rah-static--height-auto"
                  style="height: auto; overflow: visible;"
                >
                  <div>
                    <div
                      class="emotion-8"
                      data-testid="story"
                      id="blocks-core-story-plain--controls"
                    >
                      <div
                        class="story-render-container emotion-9"
                      >
                        <svg
                          aria-label="example donut graphic"
                          class="emotion-10"
                          fill="none"
                          height="128"
                          role="img"
                          stroke="currentcolor"
                          stroke-width="2"
                          viewBox="0 0 32 32"
                          width="128"
                        >
                          <circle
                            cx="16"
                            cy="16"
                            opacity="0.125"
                            r="14"
                          />
                          <circle
                            cx="16"
                            cy="16"
                            r="14"
                            stroke-dasharray="87.96459430051421"
                            stroke-dashoffset="43.982297150257104"
                            transform="rotate(-90 16 16)"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
